### ISSUE TITLE
torch.compile scatter_(value=) returns incorrect values on grouped_mm offs, MoE accumulate, Llama 4 / Qwen3 router maps, and KV-cache slots

### ISSUE THREAD

**Issue by KanakaPathivada** [torch-spyre/torch-spyre#4436]
### What happened?

`torch.compile` of `dst.scatter_(dim, index, value)` with a **Python scalar** `value` returns a tensor whose values disagree with CPU (`assert_close`, atol=rtol=0.01).

The kernel **runs** (no Inductor crash) when `torch.use_deterministic_algorithms(True)`. `dynamic=False` and `dynamic=True` match each other and both disagree with CPU. Without DeterministicGuard the same tensors die on Constant store (not this issue).

This is **`aten::scatter.value_out`**. Models do not call scalar `scatter_`; they use tensor-src `torch.scatter` / `scatter_add_`. Repro stamps a scalar onto those dest / index ranks (`torch.manual_seed(0)`).

Every **written slot** is wrong. Dest at those indices stays at the pre-scatter value (writes do not land). Not fp16 rounding. Not collision-only (unique dest `(4, 64)` is 50%).

| pattern | dest | mismatch |
|---|---|---|
| grouped_mm offs stamp `1.0` | `[128]` fp16, T=192 | **95 / 128 (74.2%)** — abs **1.0** |
| MoE accumulate stamp `1.0` | `(8, 32, 64)` fp16 | **12288 / 16384 (75.0%)** — abs **1.0** |
| Llama 4 ranks dim=1 K=1 stamp `1.0` into `-inf` | `(192, 128)` fp16 | **192 / 24576 (0.8%)** — abs **inf** (= T×K) |
| Qwen3 ranks dim=1 K=8 stamp `1.0` | `(192, 128)` fp16 | **1536 / 24576 (6.2%)** — abs **1.0** (= T×K) |
| KV-cache pad-fill slots `[3, 17]` with `0.0` | `(256, 8, 128)` bf16 | **2048 / 262144 (0.8%)** — abs **2640** (= 2×8×128) |
| YAML unique dest stamp `99.0` | `(4, 64)` fp16 | **128 / 256 (50.0%)** — abs **99.0** |

Out-of-place `torch.scatter(..., value=1.0)` on Llama 4 ranks is the same 192 / 24576.

```
Tensor-likes are not close!

Mismatched elements: 192 / 24576 (0.8%)
Greatest absolute difference: inf at index (0, 40) (up to 0.01 allowed)
```

### What did you expect to happen?

Compiled Spyre should match CPU (`assert_close`, atol=rtol=0.01).

```python
import torch

torch.manual_seed(0)
device = torch.device("spyre:0")
torch.use_deterministic_algorithms(True)

T, E, K = 192, 128, 1
dst = torch.full((T, E), float("-inf"), dtype=torch.float16)
idx = torch.zeros(T, K, dtype=torch.int64)
for t in range(T):
    idx[t] = torch.randperm(E)[:K]

def scatter_val(d, i):
    out = d.clone()
    out.scatter_(1, i, 1.0)
    return out

ref = scatter_val(dst, idx)
got = torch.compile(scatter_val, dynamic=False)(
    dst.to(device), idx.to(device)
).cpu()
torch.testing.assert_close(got, ref, atol=0.01, rtol=0.01)
```

### How can we reproduce it?

**Reproducer script:**  [scatter_scalar_mismatch.py](https://github.com/user-attachments/files/32046205/scatter_scalar_mismatch.py)

```python
import torch

torch.manual_seed(0)
device = torch.device("spyre:0")
torch.use_deterministic_algorithms(True)

T, E, K = 192, 128, 1
dst = torch.full((T, E), float("-inf"), dtype=torch.float16)
idx = torch.zeros(T, K, dtype=torch.int64)
for t in range(T):
    idx[t] = torch.randperm(E)[:K]

def scatter_val(d, i):
    out = d.clone()
    out.scatter_(1, i, 1.0)
    return out

ref = scatter_val(dst, idx)
got = torch.compile(scatter_val, dynamic=False)(
    dst.to(device), idx.to(device)
).cpu()
torch.testing.assert_close(got, ref, atol=0.01, rtol=0.01)
```

Same script: grouped_mm `[128]`, MoE `(8, 32, 64)`, Qwen3 K=8, KV-cache `(256, 8, 128)`, YAML unique `(4, 64)`.

### Any environmental details we need to know?

| | |
|-|-|
| Python | 3.12 |
| PyTorch | 2.13.0 |
| torch-spyre | commit - https://github.com/torch-spyre/torch-spyre/commit/8ef93123d6917a7b951768980b946064694d77d8 |
| Mode | `torch.compile(dynamic=False)` and `dynamic=True` with `use_deterministic_algorithms(True)` |

### Relevant log output

```Shell
python scatter_scalar_mismatch.py
cpu = expected. compiled uses assert_close (atol=rtol=0.01).
aten::scatter.value_out on model dest/index ranks. DeterministicGuard on.

--- 1. grouped_mm offs (G=128, T=192, 1-D) stamp 1.0 ---
[grouped_mm scatter_ value=1.0] compile dest (128,) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  95 / 128 (74.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 95 / 128 (74.2%)
Greatest absolute difference: 1.0 at index (0,) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0,) (up to 0.01 allowed)


[grouped_mm scatter_ value=1.0] compile_dyn dest (128,) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  95 / 128 (74.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 95 / 128 (74.2%)
Greatest absolute difference: 1.0 at index (0,) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0,) (up to 0.01 allowed)


--- 2. MoE accumulate (E=8, T=16, 32×64) stamp 1.0 ---
[MoE scatter_ value=1.0] compile dest (8, 32, 64) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  12288 / 16384 (75.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 12288 / 16384 (75.0%)
Greatest absolute difference: 1.0 at index (0, 0, 0) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.01 allowed)


[MoE scatter_ value=1.0] compile_dyn dest (8, 32, 64) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  12288 / 16384 (75.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 12288 / 16384 (75.0%)
Greatest absolute difference: 1.0 at index (0, 0, 0) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0, 0, 0) (up to 0.01 allowed)


--- 3. Llama 4 router dest (192, 128) K=1 stamp 1.0 into -inf ---
[Llama4 scatter_ value=1.0] compile dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  192 / 24576 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 192 / 24576 (0.8%)
Greatest absolute difference: inf at index (0, 40) (up to 0.01 allowed)
Greatest relative difference: inf at index (0, 40) (up to 0.01 allowed)


[Llama4 scatter_ value=1.0] compile_dyn dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  192 / 24576 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 192 / 24576 (0.8%)
Greatest absolute difference: inf at index (0, 40) (up to 0.01 allowed)
Greatest relative difference: inf at index (0, 40) (up to 0.01 allowed)


[Llama4 torch.scatter value=1.0 oop] compile dest (192, 128) torch.float16
  compile  MISMATCH  192 / 24576 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 192 / 24576 (0.8%)
Greatest absolute difference: inf at index (0, 40) (up to 0.01 allowed)
Greatest relative difference: inf at index (0, 40) (up to 0.01 allowed)


--- 4. Qwen3 router dest (192, 128) K=8 stamp 1.0 ---
[Qwen3 scatter_ value=1.0] compile dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  1536 / 24576 (6.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 1536 / 24576 (6.2%)
Greatest absolute difference: 1.0 at index (0, 3) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0, 3) (up to 0.01 allowed)


[Qwen3 scatter_ value=1.0] compile_dyn dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  1536 / 24576 (6.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 1536 / 24576 (6.2%)
Greatest absolute difference: 1.0 at index (0, 3) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0, 3) (up to 0.01 allowed)


--- 5. KV-cache dest (256, 8, 128) pad-fill slots [3, 17] ---
[KV scatter_ value=0.0] compile dest (256, 8, 128) torch.bfloat16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  2048 / 262144 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 2048 / 262144 (0.8%)
Greatest absolute difference: 2640.0 at index (17, 7, 121) (up to 0.01 allowed)
Greatest relative difference: inf at index (3, 0, 0) (up to 0.01 allowed)


[KV scatter_ value=0.0] compile_dyn dest (256, 8, 128) torch.bfloat16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  2048 / 262144 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 2048 / 262144 (0.8%)
Greatest absolute difference: 2640.0 at index (17, 7, 121) (up to 0.01 allowed)
Greatest relative difference: inf at index (3, 0, 0) (up to 0.01 allowed)


--- 6. YAML usage dest (4, 64) unique rows value=99.0 ---
[yaml scatter_ value=99.0 unique] compile dest (4, 64) torch.float16
  compile  MISMATCH  128 / 256 (50.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 128 / 256 (50.0%)
Greatest absolute difference: 99.0 at index (0, 0) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0, 0) (up to 0.01 allowed)


[yaml scatter_ value=99.0 unique] compile_dyn dest (4, 64) torch.float16
  compile_dyn  MISMATCH  128 / 256 (50.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 128 / 256 (50.0%)
Greatest absolute difference: 99.0 at index (0, 0) (up to 0.01 allowed)
Greatest relative difference: 1.0 at index (0, 0) (up to 0.01 allowed)
```
